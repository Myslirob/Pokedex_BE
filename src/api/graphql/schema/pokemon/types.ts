import { objectType } from 'nexus';

export const Attack = objectType({
    name: 'Attack',
    definition(t) {
        t.nonNull.string('name');
        t.nonNull.string('type');
        t.nonNull.int('damage');
    },
});
export const Pokemon = objectType({
    name: 'Pokemon',
    definition(t) {
        t.nonNull.id('id');
        t.nonNull.string('name');
        t.nonNull.field('weight', { type: PokemonDimension });
        t.nonNull.field('height', { type: PokemonDimension });
        t.nonNull.string('classification');
        t.nonNull.list.nonNull.string('types');
        t.nonNull.list.nonNull.string('resistant');
        t.nonNull.field('attacks', { type: PokemonAttack });
        t.nonNull.list.nonNull.string('weaknesses');
        t.nonNull.float('fleeRate');
        t.nonNull.int('maxCP');
        t.nonNull.list.nonNull.field('evolutions', { type: Pokemon });
        t.field('evolutionRequirements', { type: PokemonEvolutionRequirement });
        t.nonNull.int('maxHP');
        t.nonNull.boolean('isFavorite');
    },
});
export const PokemonAttack = objectType({
    name: 'PokemonAttack',
    definition(t) {
        t.nonNull.list.nonNull.field('fast', { type: Attack });
        t.nonNull.list.nonNull.field('special', { type: Attack });
    },
});
export const PokemonDimension = objectType({
    name: 'PokemonDimension',
    definition(t) {
        t.nonNull.string('minimum');
        t.nonNull.string('maximum');
    },
});
export const PokemonEvolutionRequirement = objectType({
    name: 'PokemonEvolutionRequirement',
    definition(t) {
        t.nonNull.int('amount');
        t.nonNull.string('name');
    },
});
