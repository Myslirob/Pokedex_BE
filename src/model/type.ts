import mongoose, { Schema } from 'mongoose';

const typeSchema = new Schema({
    name: { type: String, required: true, unique: true },
    _id: String,
});

export const Type = mongoose.model('Type', typeSchema);
