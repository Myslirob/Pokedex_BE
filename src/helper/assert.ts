export function assertNotNull<T>(value: T): asserts value is (T extends null ? never : T) {
    if (value === null) {
        throw new TypeError('Value should be not null');
    }
}
