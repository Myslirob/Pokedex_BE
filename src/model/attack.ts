import mongoose, { Schema } from 'mongoose';

export const attackSchema = new Schema({
    _id: String,
    name: { type: String, required: true, unique: true },
    type: { type: String, required: true },
    damage: { type: Number, required: true },
});

export const Attack = mongoose.model('Attack', attackSchema);

export interface AttackType {
    name: string;
    damage: number;
    type: string;
}
