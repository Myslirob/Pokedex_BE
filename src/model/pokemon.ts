import mongoose, { Schema } from 'mongoose';

export const pokemonSchema = new Schema({
    _id: String,
    name: { type: String, required: true },
    classification: { type: String, required: true },
    types: [{ type: String, ref: 'Type', required: true }],
    resistant: [{ type: String, ref: 'Type', required: true }],
    weaknesses: [{ type: String, ref: 'Type', required: true }],
    weight: {
        type: {
            minimum: { type: String, required: true },
            maximum: { type: String, required: true },
        },
        required: true,
    },
    height: {
        type: {
            minimum: { type: String, required: true },
            maximum: { type: String, required: true },
        },
        required: true,
    },
    fleeRate: { type: Number, required: true },
    evolutionRequirements: {
        type: new Schema({
            amount: { type: Number, required: true },
            name: { type: String, required: true },
        }),
        required: false,
    },
    evolutions: [{ type: String, ref: 'Pokemon' }],
    maxCP: { type: Number, required: true },
    maxHP: { type: Number, required: true },
    attacks: {
        type: {
            fast: [{ type: String, ref: 'Attack', required: true }],
            special: [{ type: String, ref: 'Attack', required: true }],
        },
        required: true,
    },
    isFavorite: { type: Boolean, default: false },
}, { _id: false });

export const Pokemon = mongoose.model('Pokemon', pokemonSchema);
