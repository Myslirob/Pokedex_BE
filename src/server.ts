import Fastify from 'fastify';
import mongoose from 'mongoose';

import { createApiRoute } from './api';

const PORT = process.env.PORT ? parseInt(process.env.PORT) : 8080;
const MONGODB = process.env.MONGODB ?? 'mongodb://root:testPassword@127.0.0.1:27017';

export async function createFastifyServer() {
    console.log('Starting...');
    const fastify = Fastify();

    console.log(`Connecting to MongoDB ${MONGODB}...`);
    await mongoose.connect(MONGODB);
    fastify.addHook('onClose', async () => {
        await mongoose.connection.close();
    });

    await createApiRoute(fastify);
    await fastify.listen({ port: PORT });
    console.log(`Server successfully started at port ${PORT}`);
    return fastify;
}
