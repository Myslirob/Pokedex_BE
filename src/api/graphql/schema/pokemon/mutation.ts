import { idArg, nonNull, objectType } from 'nexus';
import { GraphQLError } from 'graphql/error';

import { Pokemon as PokemonEntity } from '../../../../model/pokemon';

import { Pokemon } from './types';

import type { Document } from 'mongoose';
import type { Attack } from '../../../../model/attack';

export const Mutation = objectType({
    name: 'Mutation',
    definition(t) {
        t.field('favoritePokemon', {
            type: Pokemon,
            args: {
                id: nonNull(idArg()),
            },
            resolve: async (source, args) => {
                // @todo There is duplicity code via project, but I didn't have time for find correct type with populated evolution and attacks in mongoose for isolate this code to map function
                const pokemon = await PokemonEntity.findById(args.id).populate<{
                    evolutions: Document<typeof Pokemon>[];
                    attacks: { special: Document<typeof Attack>[]; fast: Document<typeof Attack>[] };
                }>(['attacks.special', 'attacks.fast', 'evolutions']).exec();
                if (pokemon === null) throw new GraphQLError(`Pokemon with id ${args.id} doesn't exist`, { extensions: { code: 'POKEMON_NOT_FOUND' } });
                pokemon.isFavorite = true;
                await pokemon.save();
                const { _id, ...data } = pokemon.toObject();
                const attacks = {
                    fast: (pokemon.attacks?.fast ?? []).map((attack) => attack.toObject()),
                    special: (pokemon.attacks?.special ?? []).map((attack) => attack.toObject()),
                };
                return {
                    ...data,
                    id: _id,
                    attacks,
                    evolutions: pokemon.evolutions.map((evolution) => {
                        const { _id, ...data } = evolution.toObject();
                        return {
                            ...data,
                            id: _id,
                        };
                    }),
                };
            },
        });
        t.field('unFavoritePokemon', {
            type: Pokemon,
            args: {
                id: nonNull(idArg()),
            },
            resolve: async (source, args) => {
                // @todo There is duplicity code via project, but I didn't have time for find correct type with populated evolution and attacks in mongoose for isolate this code to map function
                const pokemon = await PokemonEntity.findById(args.id).populate<{
                    evolutions: Document<typeof Pokemon>[];
                    attacks: { special: Document<typeof Attack>[]; fast: Document<typeof Attack>[] };
                }>(['attacks.special', 'attacks.fast', 'evolutions']).exec();
                if (pokemon === null) throw new GraphQLError(`Pokemon with id ${args.id} doesn't exist`, { extensions: { code: 'POKEMON_NOT_FOUND' } });
                pokemon.isFavorite = false;
                await pokemon.save();
                const { _id, ...data } = pokemon.toObject();
                const attacks = {
                    fast: (pokemon.attacks?.fast ?? []).map((attack) => attack.toObject()),
                    special: (pokemon.attacks?.special ?? []).map((attack) => attack.toObject()),
                };
                return {
                    ...data,
                    id: _id,
                    attacks,
                    evolutions: pokemon.evolutions.map((evolution) => {
                        const { _id, ...data } = evolution.toObject();
                        return {
                            ...data,
                            id: _id,
                        };
                    }),
                };
            },
        });
    },
});
