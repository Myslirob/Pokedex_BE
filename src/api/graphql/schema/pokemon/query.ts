import { arg, idArg, inputObjectType, nonNull, objectType, stringArg } from 'nexus';
import { GraphQLError } from 'graphql/error';

import { Type } from '../../../../model/type';
import { Pokemon as PokemonEntity } from '../../../../model/pokemon';
import { Attack } from '../../../../model/attack';

import { Pokemon } from './types';

import type { Document } from 'mongoose';

new Attack();

export const Query = objectType({
    name: 'Query',
    definition(t) {
        t.nonNull.field('pokemons', {
            type: PokemonConnection,
            args: {
                query: arg({ type: nonNull(PokemonsQueryInput) }),
            },
            resolve: async (source, { query }) => {
                // @todo There is duplicity code via project, but I didn't have time for find correct type with populated evolution and attacks in mongoose for isolate this code to map function
                const filter = {
                    ...(query.filter?.type === undefined ? {} : { types: query.filter?.type }),
                    ...(query.filter?.isFavorite === undefined ? {} : { isFavorite: query.filter.isFavorite === true }),
                    ...(query.search === undefined ? {} : { name: new RegExp(`.*${query.search}.*`, 'i') }),
                };
                const pokemons = await PokemonEntity.find(filter).sort('_id').skip(query.offset).limit(query.limit).populate<{
                    evolutions: Document<typeof Pokemon>[];
                    attacks: { special: Document<typeof Attack>[]; fast: Document<typeof Attack>[] };
                }>(['attacks.special', 'attacks.fast', 'evolutions']).exec();
                return {
                    count: pokemons.length,
                    edges: pokemons.map((pokemon) => {
                        const { _id, ...data } = pokemon.toObject();
                        const attacks = {
                            fast: (pokemon.attacks?.fast ?? []).map((attack) => attack.toObject()),
                            special: (pokemon.attacks?.special ?? []).map((attack) => attack.toObject()),
                        };
                        return {
                            ...data,
                            id: _id,
                            attacks,
                            evolutions: pokemon.evolutions.map((evolution) => {
                                const { _id, ...data } = evolution.toObject();
                                return {
                                    ...data,
                                    id: _id,
                                };
                            }),
                        };
                    }),
                    limit: query.limit,
                    offset: query.offset,
                };
            },
        });
        t.field('pokemonByName', {
            type: Pokemon,
            args: {
                name: nonNull(stringArg()),
            },
            resolve: async (source, args) => {
                // @todo There is duplicity code via project, but I didn't have time for find correct type with populated evolution and attacks in mongoose for isolate this code to map function
                const pokemon = await PokemonEntity.findOne({ name: args.name }).populate<{
                    evolutions: Document<typeof Pokemon>[];
                    attacks: { special: Document<typeof Attack>[]; fast: Document<typeof Attack>[] };
                }>(['attacks.special', 'attacks.fast', 'evolutions']).exec();
                if (pokemon === null) throw new GraphQLError(`Pokemon with name ${args.name} doesn't exist`, { extensions: { code: 'POKEMON_NOT_FOUND' } });
                const { _id, ...data } = pokemon.toObject();
                const attacks = {
                    fast: (pokemon.attacks?.fast ?? []).map((attack) => attack.toObject()),
                    special: (pokemon.attacks?.special ?? []).map((attack) => attack.toObject()),
                };
                return {
                    ...data,
                    id: _id,
                    attacks,
                    evolutions: pokemon.evolutions.map((evolution) => {
                        const { _id, ...data } = evolution.toObject();
                        return {
                            ...data,
                            id: _id,
                        };
                    }),
                };
            },
        });
        t.field('pokemonById', {
            type: Pokemon,
            args: {
                id: nonNull(idArg()),
            },
            resolve: async (source, args) => {
                // @todo There is duplicity code via project, but I didn't have time for find correct type with populated evolution and attacks in mongoose for isolate this code to map function
                const pokemon = await PokemonEntity.findById(args.id).populate<{
                    evolutions: Document<typeof Pokemon>[];
                    attacks: { special: Document<typeof Attack>[]; fast: Document<typeof Attack>[] };
                }>(['attacks.special', 'attacks.fast', 'evolutions']).exec();
                if (pokemon === null) throw new GraphQLError(`Pokemon with id ${args.id} doesn't exist`, { extensions: { code: 'POKEMON_NOT_FOUND' } });
                const { _id, ...data } = pokemon.toObject();
                const attacks = {
                    fast: (pokemon.attacks?.fast ?? []).map((attack) => attack.toObject()),
                    special: (pokemon.attacks?.special ?? []).map((attack) => attack.toObject()),
                };
                return {
                    ...data,
                    id: _id,
                    attacks,
                    evolutions: pokemon.evolutions.map((evolution) => {
                        const { _id, ...data } = evolution.toObject();
                        return {
                            ...data,
                            id: _id,
                        };
                    }),
                };
            },
        });
        t.nonNull.list.nonNull.string('pokemonTypes', {
            resolve: async () => {
                const types = await Type.find().sort('name');
                const names = types.map((type) => type.name);
                return names;
            },
        });
    },
});

export const PokemonFilterInput = inputObjectType({
    name: 'PokemonFilterInput',
    definition(t) {
        t.string('type');
        t.boolean('isFavorite');
    },
});

export const PokemonsQueryInput = inputObjectType({
    name: 'PokemonsQueryInput',
    definition(t) {
        t.nonNull.int('limit', { default: 10 });
        t.nonNull.int('offset', { default: 0 });
        t.string('search', { });
        t.field('filter', { type: PokemonFilterInput });
    },
});

const PokemonConnection = objectType({
    name: 'PokemonConnection',
    definition(t) {
        t.nonNull.int('limit');
        t.nonNull.int('offset');
        t.nonNull.int('count');
        t.nonNull.list.nonNull.field('edges', { type: Pokemon });
    },
});
