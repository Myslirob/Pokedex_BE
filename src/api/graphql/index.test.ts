import supertest from 'supertest';

import { createFastifyServer } from '../../server';
import { getNextPokemonId, toResponseObject } from '../../../test/testUtils';
import { Pokemon } from '../../model/pokemon';
import { assertNotNull } from '../../helper/assert';

import type { FastifyInstance } from 'fastify';

describe('Graphql', () => {
    let fastifyServer: Promise<FastifyInstance>;

    beforeAll(async () => {
        fastifyServer = createFastifyServer();
    });

    afterAll(async () => {
        (await fastifyServer).close();
    });

    describe('Mutation', () => {
        it('Should mark pokemon as favorite', async () => {
            const pokemon = await Pokemon.findById(getNextPokemonId());
            assertNotNull(pokemon);
            expect(pokemon.isFavorite).toBe(false);
            const mutationData = {
                query: `mutation FavoritePokemon($id: ID!) {
                  favoritePokemon(id: $id) {
                    id
                    name
                    isFavorite
                  }
                }`,
                variables: { id: pokemon.id },
            };
            const response = await supertest((await fastifyServer).server)
                .post('/api/graphql')
                .set('Content-Type', 'application/json')
                .send(mutationData);
            const responseData = toResponseObject(response);
            expect(responseData.body.data.favoritePokemon.id).toBe(pokemon.id);
            expect(responseData.body.data.favoritePokemon.name).toBe(pokemon.name);
            expect(responseData.body.data.favoritePokemon.isFavorite).toBe(true);
            const freshPokemon = await Pokemon.findById(pokemon.id).exec();
            assertNotNull(freshPokemon);
            expect(freshPokemon.isFavorite).toBe(true);
        });

        it('Should unmark pokemon as favorite', async () => {
            const pokemon = await Pokemon.findById(getNextPokemonId());
            assertNotNull(pokemon);
            pokemon.isFavorite = true;
            await pokemon.save();
            expect(pokemon.isFavorite).toBe(true);
            const mutationData = {
                query: `mutation UnfavoritePokemon($id: ID!) {
                  unFavoritePokemon(id: $id) {
                    id
                    name
                    isFavorite
                  }
                }`,
                variables: { id: pokemon.id },
            };
            const response = await supertest((await fastifyServer).server)
                .post('/api/graphql')
                .set('Content-Type', 'application/json')
                .send(mutationData);
            const responseData = toResponseObject(response);
            expect(responseData.body.data.unFavoritePokemon.id).toBe(pokemon.id);
            expect(responseData.body.data.unFavoritePokemon.name).toBe(pokemon.name);
            expect(responseData.body.data.unFavoritePokemon.isFavorite).toBe(false);
            const freshPokemon = await Pokemon.findById(pokemon.id).exec();
            assertNotNull(freshPokemon);
            expect(freshPokemon.isFavorite).toBe(false);
        });
    });

    describe('Queries', () => {
        it('Should return pokemon by id', async () => {
            const pokemon = await Pokemon.findById('001');
            assertNotNull(pokemon);

            const queryData = {
                query: `query PokemonById($pokemonByIdId: ID!) {
                  pokemonById(id: $pokemonByIdId) {
                    id
                    isFavorite
                    name
                    types
                    weaknesses
                    resistant
                    maxCP
                    maxHP
                    attacks {
                      fast {
                        name
                        damage
                      }
                    }
                    evolutions {
                      name
                      id
                    }
                  }
                }`,
                variables: { pokemonByIdId: pokemon.id },
            };
            const response = await supertest((await fastifyServer).server)
                .post('/api/graphql')
                .set('Content-Type', 'application/json')
                .send(queryData);
            const responseData = toResponseObject(response);
            expect(responseData.body.data.pokemonById.id).toBe(pokemon.id);
            expect(responseData.body.data.pokemonById.name).toBe(pokemon.name);
            expect(responseData).toMatchSnapshot({
                body: {
                    data: {
                        pokemonById: {
                            id: expect.any(String),
                            name: expect.any(String),
                            maxCP: expect.any(Number),
                            maxHP: expect.any(Number),
                            isFavorite: expect.any(Boolean),
                        },
                    },
                },
            });
        });

        it('Should return correct error when pokemon doesn\'t exist', async () => {
            const queryData = {
                query: `query PokemonById($pokemonByIdId: ID!) {
                  pokemonById(id: $pokemonByIdId) {
                    id
                    isFavorite
                    name
                    types
                    weaknesses
                    resistant
                    maxCP
                    maxHP
                  }
                }`,
                variables: { pokemonByIdId: 'xxx' },
            };
            const response = await supertest((await fastifyServer).server)
                .post('/api/graphql')
                .set('Content-Type', 'application/json')
                .send(queryData);
            const responseData = toResponseObject(response);
            expect(responseData).toMatchSnapshot();
        });

        it('Should return pokemon by name', async () => {
            const pokemon = await Pokemon.findById(getNextPokemonId());
            assertNotNull(pokemon);

            const queryData = {
                query: `query PokemonByName($name: String!) {
                   pokemonByName(name: $name) {
                    id
                    isFavorite
                    name
                    types
                    weaknesses
                    resistant
                    maxCP
                    maxHP
                  }
                }`,
                variables: { name: pokemon.name },
            };
            const response = await supertest((await fastifyServer).server)
                .post('/api/graphql')
                .set('Content-Type', 'application/json')
                .send(queryData);
            const responseData = toResponseObject(response);
            expect(responseData.body.data.pokemonByName.id).toBe(pokemon.id);
            expect(responseData.body.data.pokemonByName.name).toBe(pokemon.name);
            expect(responseData).toMatchSnapshot({
                body: {
                    data: {
                        pokemonByName: {
                            id: expect.any(String),
                            name: expect.any(String),
                            maxCP: expect.any(Number),
                            maxHP: expect.any(Number),
                            isFavorite: expect.any(Boolean),
                        },
                    },
                },
            });
        });

        describe('Should return pokemons by filter', () => {
            it('isFavorite', async () => {
                const pokemon = await Pokemon.findById(getNextPokemonId());
                assertNotNull(pokemon);
                pokemon.isFavorite = true;
                await pokemon.save();

                const queryData = {
                    query: `query Pokemons($query: PokemonsQueryInput!) {
                      pokemons(query: $query) {
                        count
                        edges {
                          classification
                          name
                          types
                          weaknesses
                          resistant
                          maxHP
                          maxCP
                          id
                          isFavorite
                        }
                        limit
                        offset
                      }
                    }`,
                    variables: { query: { limit: 20, offset: 0, filter: { isFavorite: true } } },
                };
                const response = await supertest((await fastifyServer).server)
                    .post('/api/graphql')
                    .set('Content-Type', 'application/json')
                    .send(queryData);
                const responseData = toResponseObject(response);
                expect(responseData.body.data.pokemons.edges.length).toBe(responseData.body.data.pokemons.count);
                expect(responseData.body.data.pokemons.edges.length).not.toBe(0);
                for (const pokemon of responseData.body.data.pokemons.edges) {
                    expect(pokemon.isFavorite).toBe(true);
                }
            });

            it('type', async () => {
                const queryData = {
                    query: `query Pokemons($query: PokemonsQueryInput!) {
                      pokemons(query: $query) {
                        count
                        edges {
                          classification
                          name
                          types
                          weaknesses
                          resistant
                          maxHP
                          maxCP
                          id
                          isFavorite
                        }
                        limit
                        offset
                      }
                    }`,
                    variables: { query: { limit: 20, offset: 0, filter: { type: 'Grass' } } },
                };
                const response = await supertest((await fastifyServer).server)
                    .post('/api/graphql')
                    .set('Content-Type', 'application/json')
                    .send(queryData);
                const responseData = toResponseObject(response);
                expect(responseData.body.data.pokemons.edges.length).toBe(14);
                for (const pokemon of responseData.body.data.pokemons.edges) {
                    expect(pokemon.types.includes('Grass')).toBe(true);
                }
            });

            it('type & limit & offset', async () => {
                const queryData = {
                    query: `query Pokemons($query: PokemonsQueryInput!) {
                      pokemons(query: $query) {
                        count
                        edges {
                          classification
                          name
                          types
                          weaknesses
                          resistant
                          maxHP
                          maxCP
                          id
                          isFavorite
                        }
                        limit
                        offset
                      }
                    }`,
                    variables: { query: { limit: 20, offset: 0, filter: { type: 'Poison' } } },
                };
                const response = await supertest((await fastifyServer).server)
                    .post('/api/graphql')
                    .set('Content-Type', 'application/json')
                    .send(queryData);
                const responseData = toResponseObject(response);
                expect(responseData.body.data.pokemons.edges.length).toBe(20);
                for (const pokemon of responseData.body.data.pokemons.edges) {
                    expect(pokemon.types.includes('Poison')).toBe(true);
                }

                const response2 = await supertest((await fastifyServer).server)
                    .post('/api/graphql')
                    .set('Content-Type', 'application/json')
                    .send({ ...queryData, variables: { query: { limit: 20, offset: 19, filter: { type: 'Poison' } } } });
                const responseData2 = toResponseObject(response2);
                expect(responseData.body.data.pokemons.edges[19].id).toBe(responseData2.body.data.pokemons.edges[0].id);
                expect(responseData2.body.data.pokemons.edges.length).toBe(14);
                for (const pokemon of responseData2.body.data.pokemons.edges) {
                    expect(pokemon.types.includes('Poison')).toBe(true);
                }
            });

            it('type & isFavorite', async () => {
                const pokemon = await Pokemon.findById(getNextPokemonId());
                assertNotNull(pokemon);
                pokemon.isFavorite = true;
                await pokemon.save();

                const queryData = {
                    query: `query Pokemons($query: PokemonsQueryInput!) {
                      pokemons(query: $query) {
                        count
                        edges {
                          classification
                          name
                          types
                          weaknesses
                          resistant
                          maxHP
                          maxCP
                          id
                          isFavorite
                        }
                        limit
                        offset
                      }
                    }`,
                    variables: { query: { limit: 20, offset: 0, filter: { type: pokemon.types[0], isFavorite: true } } },
                };
                const response = await supertest((await fastifyServer).server)
                    .post('/api/graphql')
                    .set('Content-Type', 'application/json')
                    .send(queryData);
                const responseData = toResponseObject(response);
                expect(responseData.body.data.pokemons.edges.length).not.toBe(0);
                for (const _pokemon of responseData.body.data.pokemons.edges) {
                    expect(_pokemon.types.includes(pokemon.types[0])).toBe(true);
                    expect(_pokemon.isFavorite).toBe(true);
                }
            });

            it('search & isFavorite & type', async () => {
                const pokemon = await Pokemon.findById(getNextPokemonId());
                assertNotNull(pokemon);
                pokemon.isFavorite = true;
                await pokemon.save();

                const searchValue = pokemon.name.slice(0, 4).toLowerCase();

                const queryData = {
                    query: `query Pokemons($query: PokemonsQueryInput!) {
                      pokemons(query: $query) {
                        count
                        edges {
                          classification
                          name
                          types
                          weaknesses
                          resistant
                          maxHP
                          maxCP
                          id
                          isFavorite
                        }
                        limit
                        offset
                      }
                    }`,
                    variables: { query: { limit: 20, offset: 0, filter: { type: pokemon.types[0], isFavorite: true }, search: searchValue } },
                };
                const response = await supertest((await fastifyServer).server)
                    .post('/api/graphql')
                    .set('Content-Type', 'application/json')
                    .send(queryData);
                const responseData = toResponseObject(response);
                expect(responseData.body.data.pokemons.edges.length).not.toBe(0);
                for (const _pokemon of responseData.body.data.pokemons.edges) {
                    expect(_pokemon.types.includes(pokemon.types[0])).toBe(true);
                    expect(_pokemon.isFavorite).toBe(true);
                    expect(_pokemon.name.toLowerCase().includes(searchValue)).toBe(true);
                }
            });
        });

        it.each([
            'PI',
            'KACH',
            'pika',
            'pikachu',
        ])('Should return Pikachu by search %p', async (searchValue) => {
            const queryData = {
                query: `query Pokemons($query: PokemonsQueryInput!) {
                  pokemons(query: $query) {
                    count
                    edges {
                      classification
                      name
                      types
                      weaknesses
                      resistant
                      maxHP
                      maxCP
                      id
                      isFavorite
                    }
                    limit
                    offset
                  }
                }`,
                variables: { query: { limit: 20, offset: 0, search: searchValue } },
            };
            const response = await supertest((await fastifyServer).server)
                .post('/api/graphql')
                .set('Content-Type', 'application/json')
                .send(queryData);
            const responseData = toResponseObject(response);
            expect(responseData.body.data.pokemons.edges.find((pokemon) => pokemon.id === '025')).not.toBe(undefined);
        });

        it('Should return empty array on search', async () => {
            const queryData = {
                query: `query Pokemons($query: PokemonsQueryInput!) {
                  pokemons(query: $query) {
                    count
                    edges {
                      classification
                      name
                      types
                      weaknesses
                      resistant
                      maxHP
                      maxCP
                      id
                      isFavorite
                    }
                    limit
                    offset
                  }
                }`,
                variables: { query: { limit: 20, offset: 0, search: 'searchValue' } },
            };
            const response = await supertest((await fastifyServer).server)
                .post('/api/graphql')
                .set('Content-Type', 'application/json')
                .send(queryData);
            const responseData = toResponseObject(response);
            expect(responseData.body.data.pokemons.edges.length).toBe(0);
        });
    });

    it('Should return pokemon types', async () => {
        const queryData = {
            query: `query Query {
                  pokemonTypes
                }`,
        };
        const response = await supertest((await fastifyServer).server)
            .post('/api/graphql')
            .set('Content-Type', 'application/json')
            .send(queryData);
        const responseData = toResponseObject(response);
        expect(responseData).toMatchSnapshot();
    });
});
