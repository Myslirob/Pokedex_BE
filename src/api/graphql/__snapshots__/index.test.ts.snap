// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graphql Queries Should return correct error when pokemon doesn't exist 1`] = `
{
  "body": {
    "data": {
      "pokemonById": null,
    },
    "errors": [
      {
        "extensions": {
          "code": "POKEMON_NOT_FOUND",
        },
        "locations": [
          {
            "column": 19,
            "line": 2,
          },
        ],
        "message": "Pokemon with id xxx doesn't exist",
        "path": [
          "pokemonById",
        ],
      },
    ],
  },
  "status": 200,
}
`;

exports[`Graphql Queries Should return pokemon by id 1`] = `
{
  "body": {
    "data": {
      "pokemonById": {
        "attacks": {
          "fast": [
            {
              "damage": 12,
              "name": "Tackle",
            },
            {
              "damage": 7,
              "name": "Vine Whip",
            },
          ],
        },
        "evolutions": [
          {
            "id": "002",
            "name": "Ivysaur",
          },
          {
            "id": "003",
            "name": "Venusaur",
          },
        ],
        "id": Any<String>,
        "isFavorite": Any<Boolean>,
        "maxCP": Any<Number>,
        "maxHP": Any<Number>,
        "name": Any<String>,
        "resistant": [
          "Water",
          "Electric",
          "Grass",
          "Fighting",
          "Fairy",
        ],
        "types": [
          "Grass",
          "Poison",
        ],
        "weaknesses": [
          "Fire",
          "Ice",
          "Flying",
          "Psychic",
        ],
      },
    },
  },
  "status": 200,
}
`;

exports[`Graphql Queries Should return pokemon by name 1`] = `
{
  "body": {
    "data": {
      "pokemonByName": {
        "id": Any<String>,
        "isFavorite": Any<Boolean>,
        "maxCP": Any<Number>,
        "maxHP": Any<Number>,
        "name": Any<String>,
        "resistant": [
          "Water",
          "Electric",
          "Grass",
          "Fighting",
          "Fairy",
        ],
        "types": [
          "Grass",
          "Poison",
        ],
        "weaknesses": [
          "Fire",
          "Ice",
          "Flying",
          "Psychic",
        ],
      },
    },
  },
  "status": 200,
}
`;

exports[`Graphql Should return pokemon types 1`] = `
{
  "body": {
    "data": {
      "pokemonTypes": [
        "Bug",
        "Dragon",
        "Electric",
        "Fairy",
        "Fighting",
        "Fire",
        "Flying",
        "Ghost",
        "Grass",
        "Ground",
        "Ice",
        "Normal",
        "Poison",
        "Psychic",
        "Rock",
        "Steel",
        "Water",
      ],
    },
  },
  "status": 200,
}
`;
